### ISSUE TITLE
2.0 Integrations API

### ISSUE THREAD

**Issue by acobster** [timber/timber#2402]
## Is your feature request related to a problem? Please describe.

**This proposal has been edited for clarity and consistency with other APIs.**

We've discussed pulling integrations (e.g. ACF) out into separate libraries. This seems like overkill to me, but I think a happy medium would be to have an official, public Integration API that we then "dogfood," implementing the current integrations in terms of that API.

This is, IMO, fairly low-effort with a big benefit for people hoping to integrate their plugins of choice.

### 1. An `IntegrationInterface`

```php
namespace Timber\Integration;

interface IntegrationInterface {
  public function should_init() : bool;
  public function init() : void;
}
```

### 2. Re-implement existing integrations in terms of this interface

Example ACF integration:

```php
use ACF;
use Timber\Integration\IntegrationInterface;

class AcfIntegration implements IntegrationInterface {
  public function should_init() : bool {
    return class_exists( ACF::class );
  }

  public function init() : void {
    add_filter('timber/post/pre_meta', [static::class, 'post_get_meta_field'], 10, 5);
    // ...all the stuff that Timber\Integrations\ACF::__construct() currently does...
  }
}
```

### 3. Register the integrations:

```php
add_filter('timber/integrations', function(array $integrations) {
  return array_merge($integrations, [
    AcfIntegration::class,
    CoAuthorsPlusIntegration::class,
    CarbonFieldsIntegration::class,
  ];
});
```

The above code would live in `Timber::init()`, where `new Integrations()` is currently called. Other theme-level code would follow the exact same pattern.

`Timber\Timber` itself then loops over the result of applying this filter, activating each one accordingly:

```php
foreach (apply_filters('timber/integrations', []) as $integration) {
  if ($integration->should_init()) {
    $integration->init();
  }
}
```

No more `Integrations` junk-drawer class; no more `maybe_init_integrations()` method. :smile: 

## Describe alternatives you’ve considered

Keeping things the same? Where's the fun in that?

## Additional context

See discussion in #2397

**Comment by gchtr** [comment]
❤️

> We've discussed pulling integrations (e.g. ACF) out into separate libraries. This seems like overkill to me, but I think a happy medium would be to have an official, public Integration API that we then "dogfood," implementing the current integrations in terms of that API.

Exactly my thoughts as well.

Having a filter for all integrations and putting the check whether an integration should be active into a function is exactly what I thought of when I thought about how integrations could be controlled better.

I have only two things that come to my mind:

Instead of `Integration::activate()`, should use `Integration::init()` as a method name, because we use `init()` more often throughout the code? If yes, we should probably also use `Integration::should_init()` instead of `Integration::should_activate()`.

And considering how Class Maps are used, would it be more consistent to use class names in the filter instead of instances?

```php
add_filter( 'timber/integrations', function( array $integrations ) {
    return array_merge( $integrations, [
        AcfIntegration::class,
        CoAuthorsPlusIntegration::class,
        CarbonFieldsIntegration::class,
    ];
} );
```

**Comment by gchtr** [comment]
Resolved in #2405.
